File: examples/conditional-policy-large/conditional-policy.R
#!/usr/bin/env Rscript

# Large synthetic conditional-policy workload.
#
# Model and data have been generated specifically for this example. Python
# and R scripts read the same MDL and CSV files so to make their results comparable.

suppressPackageStartupMessages(library(bimets))

script_arg <- grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE)
script_path <- normalizePath(sub("^--file=", "", script_arg[[1]]))
example_dir <- dirname(script_path)
model_path <- file.path(example_dir, "conditional-policy.mdl")
data_path <- file.path(example_dir, "conditional-policy-data.csv")

model_data <- CSV2BIMETS(data_path, mergedList = TRUE)
model_text <- paste(readLines(model_path, warn = FALSE), collapse = "
")

metric <- function(name, value) cat(sprintf("%s,%.6f
", name, value))

simulation_range <- c(1941, 1, 1999, 4)
exogenization <- list(y002 = c(1960, 1, 1965, 1))
adjustment_range <- window(
  model_data$y001,
  start = simulation_range[1:2],
  end = simulation_range[3:4]
)
adjustments <- list(y001 = adjustment_range * 0 + 0.02)

run <- function(simulation_type) {
  model <- LOAD_MODEL(modelText = model_text, quietly = TRUE)
  model <- LOAD_MODEL_DATA(model, model_data, quietly = TRUE)
  model <- SIMULATE(
    model,
    TSRANGE = simulation_range,
    simType = simulation_type,
    simAlgo = "NEWTON",
    simConvergence = 1e-7,
    simIterLimit = 100,
    ConstantAdjustment = adjustments,
    Exogenize = exogenization,
    JacobianDrop = "y001",
    quietly = TRUE
  )
  model
}

dynamic <- run("DYNAMIC")
static <- run("STATIC")
metric("equations", length(dynamic$vendog))
metric("dynamic_aggregate_checksum", sum(dynamic$simulation$aggregate))
metric("static_aggregate_checksum", sum(static$simulation$aggregate))
metric("dynamic_last_aggregate", tail(dynamic$simulation$aggregate, 1))

File: examples/conditional-policy-small/conditional-policy.py
#!/usr/bin/env python3
"""Small synthetic conditional-policy workload.

Model and data have been generated specifically for this example. Python
and R scripts read the same MDL and CSV files so to make their results comparable.
"""

from pathlib import Path

from bimets import (
    CSV2BIMETS,
    LOAD_MODEL,
    SIMULATE,
    BimetsDataset,
)

EXAMPLE_DIR = Path(__file__).resolve().parent
MODEL_PATH = EXAMPLE_DIR / "conditional-policy.mdl"
DATA_PATH = EXAMPLE_DIR / "conditional-policy-data.csv"


def load_data() -> BimetsDataset:
    """Load the shared quarterly CSV as a BIMETS dataset."""
    return BimetsDataset(CSV2BIMETS(DATA_PATH, merged=True))


def metric(name: str, value: float) -> None:
    """Print a stable numerical metric."""
    print(f"{name},{value:.6f}")


SIMULATION_RANGE = (2001, 1, 2019, 4)
EXOGENIZATION_RANGE = (2006, 3, 2008, 1)


def run(model, data, adjustment, simulation_type: str):
    """Run one policy scenario with conditional identities and hybrid Newton."""
    return SIMULATE(
        model,
        data,
        coefficients={},
        time_range=SIMULATION_RANGE,
        simulation_type=simulation_type,
        algorithm="NEWTON",
        convergence=1e-7,
        max_iterations=100,
        constant_adjustments={"y001": adjustment},
        exogenize={"y002": EXOGENIZATION_RANGE},
        jacobian_drop="y001",
    )


def main() -> None:
    """Compare dynamic and static conditional policy simulations."""
    model = LOAD_MODEL(model_text=MODEL_PATH.read_text(encoding="utf-8"))
    data = load_data()
    adjustment = (
        data["y001"][[list(SIMULATION_RANGE[:2]), list(SIMULATION_RANGE[2:])]] * 0
        + 0.02
    )
    dynamic = run(model, data, adjustment, "DYNAMIC")
    static = run(model, data, adjustment, "STATIC")
    metric("equations", len(model.endogenous))
    metric("dynamic_aggregate_checksum", dynamic["aggregate"].values.sum())
    metric("static_aggregate_checksum", static["aggregate"].values.sum())
    metric("dynamic_last_aggregate", dynamic["aggregate"].values[-1])


if __name__ == "__main__":
    main()
